Backport 2.28: Copy files instead of hard-linking on Windows (#6744)
Mbed-TLS/mbedtls: CMakeLists.txt

# Create a symbolic link from ${base_name} in the binary directory
# to the corresponding path in the source directory.
# Note: Copies the file(s) on Windows.
function(link_to_source base_name)
    set(link "${CMAKE_CURRENT_BINARY_DIR}/${base_name}")
    set(target "${CMAKE_CURRENT_SOURCE_DIR}/${base_name}")

    if (NOT EXISTS ${link})
        if (CMAKE_HOST_UNIX)
            execute_process(COMMAND ln -s ${target} ${link}
                RESULT_VARIABLE result
                ERROR_VARIABLE output)

            if (NOT ${result} EQUAL 0)
                message(FATAL_ERROR "Could not create symbolic link for: ${target} --> ${output}")
            endif()
        else()
            if (IS_DIRECTORY ${target})

>                 file(GLOB_RECURSE files FOLLOW_SYMLINKS LIST_DIRECTORIES false RELATIVE ${target} "${target}/*")

Reviewer: `file(GLOB_RECURSE` does not have the `LIST_DIRECTORIES` option in cmake 2.28.12, which we must support in 2.28.

Author: What can we use instead? I don't know cmake. I can test if I manage to find a cmake 2.28.12 executable for Windows.

Reviewer: `LIST_DIRECTORIES false` filters out directories from the list. We can manually check whether each file is a directory or not to predicate whether we link it. I'll put a suggestion shortly.

Author: I don't know what's going on, but the code as it stands now is working with `cmake version 2.8.12.2`.

Author: If I undesrtand correctly, in CMake 2.8:

* [`file(GLOB_RECURSE …)`](https://cmake.org/cmake/help/v2.8.12/cmake.html#command:file) does not recognize `LIST_DIRECTORIES` so it's treating `LIST_DIRECTORIES` and `false` as extra glob patterns to match, which makes no difference. (And maybe it's also ignoring `RELATIVE ${target}`, but that only mattered when symlinking, not when copying the files.)
* Directories are passed to [`configure_file`](https://cmake.org/cmake/help/v2.8.12/cmake.html#command:configure_file), which causes CMake to know about the directories if they're ever referenced as targets, but they're not.

So it's working by accident, but might cause someone to hate us if they need to do further debugging on it.

Reviewer: That's correct I believe. The documentation for `configure_file()` says:
> The <input> must be a file, not a directory.

Which is why the original PR filtered out directories with `LIST_DIRECTORIES`. I'm not sure why this doesn't cause a problem. Perhaps `configure_file()` in 2.8.12 either doesn't care in practice or just ignores directories.

The `RELATIVE` flag also exists in 2.8.12, so that shouldn't be a problem.

>                 file(GLOB_RECURSE files FOLLOW_SYMLINKS LIST_DIRECTORIES false RELATIVE ${target} "${target}/*")
>                 foreach(file IN LISTS files)
>                     configure_file("${target}/${file}" "${link}/${file}" COPYONLY)
>                 endforeach(file)

Reviewer: Draft
```suggestion
                file(GLOB_RECURSE files FOLLOW_SYMLINKS RELATIVE ${target} "${target}/*")
                foreach(file IN LISTS files)
                    if(NOT IS_DIRECTORY ${file})
                        configure_file("${target}/${file}" "${link}/${file}" COPYONLY)
                    endif()
                endforeach(file)
```

Author: I've applied this change and I checked that it has the same effect with CMake 2.8.12 on Windows 10.
